proposals/20220823-huggingface_model_pusher.md

```diff
@@ -0,0 +1,73 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+---
+
+**Your name:** Chansung Park
+
+**Your email:** deep.diver.csp@gmail.com
+
+**Your company/organization:** Individual(ML GDE)
+
+**Project name:** HuggingFace Model Pusher
+
+## Project Description
+HuggingFace Model Pusher(`HFModelPusher`) pushes blessed model to the [HuggingFace Model Hub](https://huggingface.co/models).
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+HuggingFace Model Hub let us have [Git-LFS](https://git-lfs.github.com) enabled repositories in public and private. Some model types hosted in HuggingFace Model Hub can be directly loaded/used with APIs of [transformers](https://huggingface.co/docs/transformers/index) package. However, it is not limited. We can host arbitrary types of model too. 
+
+Plus, HuggingFace Model Hub is easy to manage model versions, especially who is familar with Git.
+
+## Project Implementation
+HFModelPusher is a class-based TFX component, and it inherits from TFX standard `Pusher` component.
+
+It takes the following inputs:
+```
+HFModelPusher(
+    username: str,
+    huggingface_access_token: str,
+    repo_name: Optional[str],    
+    model: Optional[types.Channel] = None,
+    model_blessing: Optional[types.Channel] = None,    
+)
+```
+- `username` : the username of the HuggingFace Hub
+- `huggingface_access_token` : access token value of the HuggingFace Hug user. 
+- `repo_name` : the repository name to push the current version of the model to. The default value is same as the TFX pipeline name
+- `model` : the model artifact from the upstream TFX component such as `Trainer`
+- `model_blessing` : the blessing artifact from the upstream TFX component such as `Evaluator`
+
+It gives the follwing outputs:
+- `pushed` : integer value to denote if the model is pushed or not. This is set to 0 when the input model is not blessed, and it is set to 1 when the model is successfully pushed
+- `pushed_version` : string value to indicate the current model version. This is decided by `time.time()` Python built-in function
+- `repo_id` : repository ID where the model is pushed to. This follows the format of f"{username}/{repo_name}"
+- `branch` : branch name where the model is pushed to. The branch name is automatically assigned to the same value of  `pushed_version`
+- `repo_url` : repository URL. It is something like f"https://huggingface.co/{repo_id}/{branch}"
+
+The behaviour of the component:
+1. It pushes the model when the model is blessed or when the `model_blessing` parameter is set to `None`
+2. Create HuggingFace Repository object using `huggingface-hub` package. It will clone one if there is already existing repository.
+3. Checkout a new branch with the name as `pushed_version`.
```

Comment:
If it would check out a new branch, how is the end user supposed to use the model using `from_pretrained()`? 

I know HF Hub can be used as a generic tool for hosting models of any type. But for an HF user, it would be a natural expectation to use the model checkpoints pushed to the Hub using APIs like `from_pretrained()`. We need to clarify this. 

How will we handle the README.md of the underlying Hub repository? Here's how HF does it: 

https://github.com/huggingface/transformers/blob/f0d496828d3da3bf1e3c8fbed394d7847e839fa6/src/transformers/trainer.py#L3274

Reply:
> If it would check out a new branch, how is the end user supposed to use the model using from_pretrained()?

Since the model is pushed for experimental purpose, I think it would be good to track the versions of the model within separate branches. When the model is ready to be open to public, one can manually merge the right version(branch) into the `main` branch. 

> How will we handle the README.md of the underlying Hub repository? Here's how HF does it:

I have not thought about the model card. Should we need it? It looks like `create_model_card()` is useful if the model has some dependencies to the other HuggingFace features such as `Datasets`

Maybe we can prepare a template `README.md`, then inject some values into it, then add it to the commit. Or we can add this feature after the #82 (Model Card Generator) is fully developed. WDYT?

Reply:
> Since the model is pushed for experimental purpose, I think it would be good to track the versions of the model within separate branches. When the model is ready to be open to public, one can manually merge the right version(branch) into the main branch.

Let's add this info to the proposal. 

> I have not thought about the model card. Should we need it? It looks like create_model_card() is useful if the model has some dependencies to the other HuggingFace features such as Datasets

Maybe we can prepare a template README.md, then inject some values into it, then add it to the commit. Or we can add this feature after the https://github.com/tensorflow/tfx-addons/issues/82 (Model Card Generator) is fully developed. WDYT?

All good.

Reply:
> Let's add this info to the proposal.

Absolutely! Thanks for the good one